Feature/moderation hallucination eval multilingual translation

SnowMasaya opened the pull request:
## Description

# feat: Add multilingual translation support for evaluation pipeline(moderation and hallucination)

## 📋 Summary

This PR introduces comprehensive multilingual translation capabilities to the NeMo-Guardrails evaluation system, enabling users to evaluate AI models across different languages and cultures. The implementation includes a flexible translation provider system, caching mechanisms, and seamless integration with existing evaluation workflows.


## 🚀 Key Features

### 🌍 Multilingual Translation System
- **Flexible Translation Providers**: Support for both local (HuggingFace) and remote (DeepL, NVIDIA Riva) translation services
- **Translation Caching**: Intelligent caching system to avoid redundant translations and improve performance
- **Configurable Backends**: Easy configuration for different translation services via YAML configs
- **Progress Tracking**: Real-time progress bars for translation operations

## Dependencies Added

- `deepl` (^1.22.0) - DeepL translation service integration
- `nvidia-riva-client` (^2.21.0) - NVIDIA Riva translation service
- `torch` (^2.7.1) - PyTorch for local translation models
- `transformers` (^4.53.0) - HuggingFace transformers for local models
- `sentencepiece` (^0.2.0) - Tokenization support

##  Architecture

### Translation Provider System

```
nemoguardrails/evaluate/langproviders/
├── base.py # Base provider interface
├── local.py # HuggingFace-based local translator
├── remote.py # Remote service providers (DeepL, Riva)
├── configs/ # Translation service configurations
└── README.md # Provider documentation
```

### Core Components
- **`utils_translate.py`**: Core translation utilities and caching
- **Enhanced `utils.py`**: Integration with dataset loading
- **Updated evaluation modules**: Multilingual support in hallucination and moderation evaluation
- **CLI enhancements**: Translation configuration support

## 🔄 Usage Examples

### Basic Translation Configuration
```yaml
# translation.yaml
provider: "deepl"
api_key: "${DEEPL_API_KEY}"
target_language: "ja"
```

### CLI Usage
```bash
# Evaluate with translation
nemoguardrails evaluate hallucination \
  --dataset data/hallucination/sample.txt \
  --translation-config configs/translation.yaml

# Evaluate moderation with Japanese translation
nemoguardrails evaluate moderation \
  --dataset data/moderation/harmful.txt \
  --translation-config configs/japanese_translation.yaml
```

## 🧪 Testing

The implementation includes comprehensive test coverage:

- **Provider Tests**: Unit tests for all translation providers
- **Integration Tests**: End-to-end translation workflow testing
- **Cache Tests**: Translation caching mechanism validation
- **CLI Tests**: Command-line interface testing with translation support

Run tests with:
```bash
pytest tests/eval/translate/ -v
```

## Configuration

### Translation Service Configuration
```yaml
# DeepL Configuration
provider: "deepl"
api_key: "${DEEPL_API_KEY}"
target_language: "ja"

# HuggingFace Local Configuration
provider: "huggingface"
model_name: "Helsinki-NLP/opus-mt-en-ja"
target_language: "ja"
device: "cpu"

# NVIDIA Riva Configuration
provider: "riva"
url: "https://riva-server:8000"
target_language: "ja"
```

## Breaking Changes

None. This is a purely additive feature that maintains full backward compatibility.

## 📝 Documentation

- Added comprehensive README for translation providers
- Updated evaluation documentation with multilingual examples
- Added configuration examples for all supported translation services

## 🎯 Impact

This enhancement significantly expands NeMo-Guardrails' evaluation capabilities, making it a truly global tool for AI safety and compliance evaluation across different languages and cultures.

## Checklist

- [x] I've read the [CONTRIBUTING](https://github.com/NVIDIA/NeMo-Guardrails/blob/develop/CONTRIBUTING.md) guidelines.
- [x] I've updated the documentation if applicable.
- [x] I've added tests if applicable.
- [ ] @mentions of the person or team responsible for reviewing proposed changes.

masayaOgushi committed:
feat: add multilingual translation dependencies

masayaOgushi committed:
feat: implement multilingual translation system

masayaOgushi committed:
feat: integrate multilingual support in evaluation pipeline

masayaOgushi committed:
test: add comprehensive translation system tests

masayaOgushi committed:
fix: add langchain-nvidia-ai-endpoint, remove pyproject-toml

masayaOgushi committed:
fix: copilot advice base

masayaOgushi committed:
fix: remove extra test

masayaOgushi committed:
fix: multilingual translation dependencies

masayaOgushi committed:
fix: make pre_commit related issues

trebedea reviewed (changes requested):
Provided several comments - most are just nice to have. 
@SnowMasaya try to fix the ones you feel are most important - e.g. some duplicated code and documentation related. 

@Pouyanpi can you check if you have any feedback related to tests and using poetry?

SnowMasaya commented on nemoguardrails/evaluate/langproviders/base.py:
```diff
@@ -0,0 +1,99 @@
+                )
+
+            # Validate environment variable and set API key before loading the provider
+            if hasattr(self, "ENV_VAR"):
```
I add comments on the README.md

SnowMasaya commented on nemoguardrails/evaluate/langproviders/local.py:
```diff
@@ -0,0 +1,169 @@
+    """
+
+    DEFAULT_PARAMS = {
+        "model_name": "Helsinki-NLP/opus-mt-{}",  # This is inconsistent with generators and may change to `name`.
```
I will remove this comment. Thank you for mention

SnowMasaya commented on nemoguardrails/evaluate/langproviders/local.py:
```diff
@@ -0,0 +1,169 @@
+                return "cuda"
+            else:
+                return "cpu"
+        except ImportError:
```
This is incorrect Exception. I will change it.

SnowMasaya commented on nemoguardrails/evaluate/langproviders/local.py:
```diff
@@ -0,0 +1,169 @@
+            self.target_lang = self.lang_overrides.get(
+                self.target_lang, self.target_lang
+            )
+            model_suffix = f"{self.source_lang}-{self.target_lang}"
```
OK, I add it to the doc.

SnowMasaya commented on nemoguardrails/evaluate/langproviders/remote.py:
```diff
@@ -0,0 +1,185 @@
+
+    # fmt: off
+    # Reference: https://docs.nvidia.com/nim/riva/nmt/latest/support-matrix.html#models
+    lang_support = [
```
It would be better but it lead to become unstable code if the web site format change.

SnowMasaya commented on nemoguardrails/evaluate/langproviders/remote.py:
```diff
@@ -0,0 +1,185 @@
+    }
+
+    def _load_langprovider(self):
+        from deepl import Translator
```
OK. I add this implementation.

SnowMasaya commented on nemoguardrails/evaluate/langproviders/remote.py:
```diff
@@ -0,0 +1,185 @@
+            raise Exception(
+                f"Language pair {self.language} is not supported for {self.__class__.__name__} services."
+            )
+        self._source_lang = self.source_lang
```
DeepL does not need lang_overrides for source code within my observation range.

SnowMasaya commented on nemoguardrails/evaluate/utils_translate.py:
```diff
@@ -0,0 +1,239 @@
+        """Generate cache key from text and target language."""
+        # Create a hash of the text and target language
+        content = f"{text}:{target_lang}"
+        return content
```
OK, I change it.

SnowMasaya commented on nemoguardrails/evaluate/langproviders/README.md:
```diff
@@ -0,0 +1,392 @@
+Supports 100 languages (see the [official documentation](https://huggingface.co/facebook/m2m100_418M#languages-covered) for details)
+
+### DeeplTranslator
+Supports 29 languages:
```
OK

SnowMasaya commented on nemoguardrails/evaluate/langproviders/base.py:
```diff
@@ -0,0 +1,99 @@
+# SPDX-License-Identifier: Apache-2.0
+
+
+"""Translator that translates a prompt."""
```
OK

SnowMasaya commented on nemoguardrails/evaluate/langproviders/base.py:
```diff
@@ -0,0 +1,99 @@
+from typing import List
+
+
+class LangProvider:
```
OK

SnowMasaya commented on nemoguardrails/evaluate/utils_translate.py:
```diff
@@ -0,0 +1,262 @@
+            if isinstance(item, dict):
+                # For JSON format, translate specific fields
+                translated_item = item.copy()
+                for field in ["answer", "question", "evidence"]:
```
OK

SnowMasaya commented on nemoguardrails/evaluate/utils_translate.py:
```diff
@@ -0,0 +1,262 @@
+                # For text format
+                original_text = item.strip()
+                # Check cache first
+                cached_translation = cache.get(original_text, translator.target_lang)
+                if cached_translation:
+                    translated_dataset.append(cached_translation)
+                else:
+                    # Translate and cache
+                    translated_text = translator._translate(original_text)
+                    translated_dataset.append(translated_text)
+                    cache.set(original_text, translator.target_lang, translated_text)
```
OK

SnowMasaya commented on pyproject.toml:
```diff
@@ -101,6 +101,12 @@ google-cloud-language = { version = ">=2.14.0", optional = true }
+nvidia-riva-client = "^2.21.0"
+torch = "^2.7.1"
+transformers = "^4.53.0"
+sentencepiece = "^0.2.0"
```
Do you mean changing the following like this?

```
[tool.poetry.extras]
:
:
eval = ["tqdm", "numpy", "streamlit", "tornado",  'deepl = "^1.22.0"', 'nvidia-riva-client = "^2.21.0"', 'torch = "^2.7.1"', 'transformers = "^4.53.0"', 'sentencepiece = "^0.2.0"',]
:
:
all = [
  "presidio-analyzer",
  "presidio-anonymizer",
  "tqdm",
  "numpy",
  "streamlit",
  "langchain-openai",
  "google-cloud-language",
  "opentelemetry-api",
  "opentelemetry-sdk",
  "aiofiles",
  "langchain-nvidia-ai-endpoints",
  "yara-python",
  'deepl = "^1.22.0"',
  'nvidia-riva-client = "^2.21.0"',
  'torch = "^2.7.1"',
  'transformers = "^4.53.0"',
  'sentencepiece = "^0.2.0"',
]
```

trebedea replied:
I was thinking to have a new dependency `translation` or `eval-translation` with the extra dependencies required for translation.

Pouyanpi replied:
I'll take care of the dependencies 👍🏻

masayaOgushi committed:
fix: Remove redundant test files

masayaOgushi committed:
fix: add wanring, common process

masayaOgushi committed:
fix: add configurable endpoint support for RivaTranslator and refactor translation code

- Add YAML configurable endpoints to RivaTranslator (remote.py):
  * Support uri parameters from YAML config
  * Local mode: only uri can be overridden, others use defaults

- Refactor translation utilities (utils_translate.py):
  * Extract _check_cache_and_translate() helper function
  * Eliminate duplicate cache checking and translation logic
  * Simplify load_dataset() function while preserving functionality
  * Reduce code duplication across different file formats

- Update translation provider tests (base.py, local.py):
  * Fix test configurations to use list format for langproviders
  * Remove assertions on non-existent attributes
  * Update error handling for new validation logic
  * Ensure compatibility with configurable endpoint feature

masayaOgushi committed:
fix: update translation tests for configurable endpoint feature

- Fix test configurations to use list format for langproviders
- Remove obsolete assertions on non-existent attributes
- Add configurable endpoint tests to test_remote_translators.py
- Update cache tests to work with new translation logic
- Consolidate RivaTranslator tests in single file

masayaOgushi committed:
docs: add configurable endpoint examples to langproviders README

- Add YAML examples for RivaTranslator endpoint configuration
- Document local  mode parameter behavior
- Update existing examples for consistency

Helps users configure RivaTranslator endpoints via YAML.

masayaOgushi committed:
fix: add white space

masayaOgushi committed:
fix: None value case support

trebedea commented on nemoguardrails/evaluate/langproviders/local.py:
```diff
@@ -0,0 +1,169 @@
+    def _select_hf_device(self):
+        """Select the appropriate device for HuggingFace models."""
+        try:
+            if torch.cuda.is_available():
```
`hf_args` still appers in the README. Should it be removed if we don't use it?

masayaOgushi committed:
Fix: README, pyptoject.toml

- README: remove hf_args
- pyproject.toml: update dependency for translation

masayaOgushi committed:
# Update
standardize text normalization across evaluation modules

- Rename normalize_check to normalize_text for better semantic clarity
- Simplify text normalization logic to handle hallucination_agreement values
- Integrate normalize_text function in hallucination evaluation
- add dataset name to tranclation cache file

masayaOgushi committed:
# fix test code

- change cache file name

masayaOgushi committed:
# Add eiva config yaml

add example setting yaml
